python/tchannel/tornado/stream.py

```
@@ -0,0 +1,143 @@
+    def write(self, chunk):
+        if self._state == StreamState.completed:
+            raise StreamingException("Stream has been closed.")
+
+        self._stream += chunk
+        self._condition.notify()
+
+    @tornado.gen.coroutine
+    def close(self):
+        self._state = StreamState.completed
+        self._condition.notify()
+
+
+class PipeArgStream(ArgStream):
+
+    def __init__(self, rpipe, wpipe=None, auto_close=False):
+        """Pipe-based arg stream
+
+        NOTE: reading from or writing to files, use os.open to get the file
+        descriptor instead of python's open. Socket file descriptors and
+        others are fine.
+
+        when you use os.pipe to generate one write pipe and one read pipe, you
+        need to pass both of them into init method.
+
+        :param rpipe: an integer file descriptor which supports read ops
+        :param wpipe: an integer file descriptor which supports write ops
+        :param auto: flag to indicate to close the stream automatically or not
+        """
+        super(PipeArgStream, self).__init__()
+        self._rpipe = rpipe
+        self._wpipe = wpipe
+
+        self._rs = (PipeIOStream(self._rpipe) if
+                    self._rpipe is not None else None)
+        self._ws = (PipeIOStream(self._wpipe) if
+                    self._wpipe is not None else None)
+        self.auto_close = auto_close
+
+    @tornado.gen.coroutine
+    def read(self):
+        if self._state == StreamState.completed or self._rpipe is None:
+            raise tornado.gen.Return("")
+
+        chunk = ""
+        try:
+            chunk = yield self._rs.read_bytes(
+                common.MAX_PAYLOAD_SIZE, partial=True)
+
+        except StreamClosedError:
+            # reach the end of the pipe stream
+            self._state = StreamState.completed
+        finally:
+            raise tornado.gen.Return(chunk)
+
+    @tornado.gen.coroutine
+    def write(self, chunk):
+        if self._wpipe is not None:
```

This should be an assert instead of a silent failure